File: src/java/foundgine-providers/src/main/java/com/foundgine/providers/storage/elasticsearch/ElasticsearchSemanticLexicalCandidateSource.java
package com.foundgine.providers.storage.elasticsearch;

import com.fasterxml.jackson.databind.*;
import com.foundgine.core.abstractions.*;
import com.foundgine.core.semantic.resolution.*;
import java.net.*;
import java.net.http.*;
import java.util.*;

/**
 * Elasticsearch lexical retrieval using the REST API; context is a retrieval
 * hint only.
 */
public final class ElasticsearchSemanticLexicalCandidateSource implements ISemanticLexicalCandidateSource {
	private final HttpClient client;
	private final URI base;
	private final String index;
	private final ObjectMapper mapper;

	public ElasticsearchSemanticLexicalCandidateSource(HttpClient c, URI base, String index) {
		client = Objects.requireNonNull(c);
		this.base = Objects.requireNonNull(base);
		this.index = Objects.requireNonNull(index);
		mapper = new ObjectMapper();
	}

	@Override
	public List<SemanticLexicalCandidate> retrieve(SemanticLexicalRequest r) {
		try {
			Map<String, Object> mm = Map.of("query", r.token(), "fields",
					List.of("canonicalName^4", "aliases^3", "searchText^2", "description"), "fuzziness", "AUTO");
			Map<String, Object> body = new LinkedHashMap<>();
			body.put("size", r.limit());
			body.put("query", Map.of("bool", Map.of("must", List.of(Map.of("multi_match", mm)), "filter", List
					.of(Map.of("terms", Map.of("kind", r.effectiveKinds().stream().map(Enum::toString).toList()))))));
			HttpRequest req = HttpRequest.newBuilder(base.resolve("/" + URLEncoder.encode(index, "UTF-8") + "/_search"))
					.header("Content-Type", "application/json")
					.POST(HttpRequest.BodyPublishers.ofString(mapper.writeValueAsString(body))).build();
			HttpResponse<String> res = client.send(req, HttpResponse.BodyHandlers.ofString());
			if (res.statusCode() / 100 != 2)
				throw new IllegalStateException("Elasticsearch returned " + res.statusCode());
			JsonNode hits = mapper.readTree(res.body()).path("hits").path("hits");
			List<SemanticLexicalCandidate> out = new ArrayList<>();
			for (JsonNode h : hits) {
				JsonNode s = h.path("_source");
				if (s.path("canonicalName").isMissingNode())
					continue;
				SemanticLexicalCandidateKind k = SemanticLexicalCandidateKind
						.valueOf(s.path("kind").asText().toUpperCase(Locale.ROOT));
				double score = h.path("_score").asDouble();
				out.add(new SemanticLexicalCandidate(r.token(), k, s.path("canonicalName").asText(), score,
						uid(s, "entityId"), rid(s, "relationshipId"), fid(s, "fieldId"), uid(s, "sourceEntityId"),
						uid(s, "targetEntityId"), s.path("value").isMissingNode() ? null : s.path("value").asText(),
						List.of(new ResolutionEvidence("Elasticsearch lexical match", CandidateEvidenceKind.BM25,
								score))));
			}
			return out;
		} catch (Exception e) {
			if (e instanceof RuntimeException x)
				throw x;
			throw new IllegalStateException("Elasticsearch retrieval failed", e);
		}
	}

	private static EntityId uid(JsonNode n, String x) {
		return n.hasNonNull(x) ? new EntityId(n.get(x).asLong()) : null;
	}

	private static RelationshipId rid(JsonNode n, String x) {
		return n.hasNonNull(x) ? new RelationshipId(n.get(x).asLong()) : null;
	}

	private static FieldId fid(JsonNode n, String x) {
		return n.hasNonNull(x) ? new FieldId(n.get(x).asLong()) : null;
	}
}

File: src/csharp/Foundgine.Providers/Storage/Elasticsearch/ElasticsearchSemanticLexicalCandidateSource.cs
using System.Net.Http.Json;
using System.Text.Json;
using Foundgine.Core.Abstractions;
using Foundgine.Core.Semantic.Resolution;

namespace Foundgine.Providers.Storage.Elasticsearch;

/// <summary>
/// Elasticsearch implementation of Foundgine's provider-neutral lexical
/// candidate source. Elasticsearch supplies ranked hypotheses; Foundgine still
/// validates semantic topology and produces the final interpretation.
/// </summary>
public sealed class ElasticsearchSemanticLexicalCandidateSource : ISemanticLexicalCandidateSource
{
    private readonly HttpClient _httpClient;
    private readonly string _index;
    private readonly JsonSerializerOptions _jsonOptions = new(JsonSerializerDefaults.Web);

    public ElasticsearchSemanticLexicalCandidateSource(
        HttpClient httpClient,
        string index = "foundgine-semantic-lexicon")
    {
        _httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
        if (string.IsNullOrWhiteSpace(index))
            throw new ArgumentException("Elasticsearch index cannot be empty.", nameof(index));
        _index = index;
    }

    public IReadOnlyList<SemanticLexicalCandidate> Retrieve(SemanticLexicalRequest request)
    {
        ArgumentNullException.ThrowIfNull(request);
        return RetrieveAsync(request, CancellationToken.None).GetAwaiter().GetResult();
    }

    public IReadOnlyList<SemanticLexicalCandidate> Retrieve(
        SemanticLexicalRequest request,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(request);
        return RetrieveAsync(request, cancellationToken).GetAwaiter().GetResult();
    }

    public async Task<IReadOnlyList<SemanticLexicalCandidate>> RetrieveAsync(
        SemanticLexicalRequest request,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);

        var kindNames = request.EffectiveKinds.Select(x => x.ToString()).ToArray();
        var body = new
        {
            size = request.Limit,
            query = new
            {
                @bool = new
                {
                    must = new object[]
                    {
                        new
                        {
                            multi_match = new
                            {
                                query = request.Token,
                                fields = new[] { "canonicalName^4", "aliases^3", "searchText^2", "description" },
                                fuzziness = "AUTO"
                            }
                        }
                    },
                    filter = new object[]
                    {
                        new { terms = new { kind = kindNames } }
                    }
                }
            }
        };

        if (request.ContextEntity is not null)
        {
            // Context is deliberately a retrieval hint, not a semantic
            // authorization check. The core resolver performs authoritative
            // graph compatibility validation.
            body = new
            {
                size = request.Limit,
                query = new
                {
                    @bool = new
                    {
                        must = new object[]
                        {
                            new
                            {
                                multi_match = new
                                {
                                    query = request.Token,
                                    fields = new[] { "canonicalName^4", "aliases^3", "searchText^2", "description" },
                                    fuzziness = "AUTO"
                                }
                            }
                        },
                        filter = new object[]
                        {
                            new { terms = new { kind = kindNames } },
                            new
                            {
                                @bool = new
                                {
                                    should = new object[]
                                    {
                                        new { term = new { entityId = request.ContextEntity.Value.Value } },
                                        new { term = new { sourceEntityId = request.ContextEntity.Value.Value } },
                                        new { term = new { targetEntityId = request.ContextEntity.Value.Value } }
                                    },
                                    minimum_should_match = 1
                                }
                            }
                        }
                    }
                }
            };
        }

        using var response = await _httpClient.PostAsJsonAsync($"/{Uri.EscapeDataString(_index)}/_search", body,
            _jsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        await using var stream = await response.Content.ReadAsStreamAsync(cancellationToken);
        using var document = await JsonDocument.ParseAsync(stream, cancellationToken: cancellationToken);

        if (!document.RootElement.TryGetProperty("hits", out var hits) ||
            !hits.TryGetProperty("hits", out var hitArray))
            return [];

        var results = new List<SemanticLexicalCandidate>();
        foreach (var hit in hitArray.EnumerateArray())
        {
            if (!hit.TryGetProperty("_source", out var source)) continue;
            var candidate = ReadCandidate(request.Token, hit, source);
            if (candidate is not null) results.Add(candidate);
        }

        return results
            .OrderByDescending(x => x.Score)
            .ThenBy(x => x.CanonicalName, StringComparer.OrdinalIgnoreCase)
            .Take(request.Limit)
            .ToArray();
    }

    private static SemanticLexicalCandidate? ReadCandidate(
        string token,
        JsonElement hit,
        JsonElement source)
    {
        if (!source.TryGetProperty("kind", out var kindProperty) ||
            !Enum.TryParse<SemanticLexicalCandidateKind>(kindProperty.GetString(), true, out var kind) ||
            !source.TryGetProperty("canonicalName", out var canonicalProperty))
            return null;

        var score = hit.TryGetProperty("_score", out var scoreProperty) &&
                    scoreProperty.ValueKind == JsonValueKind.Number
            ? scoreProperty.GetDouble()
            : 0d;

        return new SemanticLexicalCandidate(
            token,
            kind,
            canonicalProperty.GetString() ?? string.Empty,
            score,
            ReadEntityId(source, "entityId"),
            ReadRelationshipId(source, "relationshipId"),
            ReadFieldId(source, "fieldId"),
            ReadEntityId(source, "sourceEntityId"),
            ReadEntityId(source, "targetEntityId"),
            source.TryGetProperty("value", out var value) ? value.GetString() : null,
            [
                new ResolutionEvidence(
                    $"Elasticsearch lexical match for '{token}'.",
                    CandidateEvidenceKind.Bm25,
                    score)
            ]);
    }

    private static EntityId? ReadEntityId(JsonElement source, string name) =>
        source.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.Number
            ? new EntityId(value.GetUInt64())
            : null;

    private static RelationshipId? ReadRelationshipId(JsonElement source, string name) =>
        source.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.Number
            ? new RelationshipId(value.GetUInt64())
            : null;

    private static FieldId? ReadFieldId(JsonElement source, string name) =>
        source.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.Number
            ? new FieldId(value.GetUInt64())
            : null;
}